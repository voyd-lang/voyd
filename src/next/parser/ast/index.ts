export * from "./expr.js";
export * from "./form.js";
export * from "./atom.js";
export * from "../reader-macros/lib/init-helpers.js";
export { is, SourceLocation } from "./syntax.js";
