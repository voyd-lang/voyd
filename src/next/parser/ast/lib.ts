import { FloatAtom, IdentifierAtom, IntAtom } from "./atom.js";
import { Expr } from "./expr.js";
import {
  ArrayLiteralForm,
  CallForm,
  FormElementInitVal,
  LabelForm,
  ObjectLiteralForm,
  ParenForm,
  TupleForm,
} from "./form.js";
import { is } from "./index.js";

export const call = (
  fn: IdentifierAtom | string,
  ...args: FormElementInitVal
) => new CallForm([fn, ...args]);

export const paren = (...args: FormElementInitVal) => new ParenForm(args);

export const tuple = (...args: FormElementInitVal) => new TupleForm(args);

export const arrayLiteral = (...args: FormElementInitVal) =>
  new ArrayLiteralForm(args);

export const objectLiteral = (...args: LabelForm[]) =>
  new ObjectLiteralForm(args);

export const label = (label: IdentifierAtom | string, value: Expr) =>
  new LabelForm([label, value]);

export const identifier = (id: string) => new IdentifierAtom(id);

export const int = (value: string | number, type: "i32" | "i64" = "i32") =>
  new IntAtom(String(value)).setType(type);

export const float = (value: string | number, type: "f32" | "f64" = "f64") =>
  new FloatAtom(String(value)).setType(type);

export const string = (value: string) => {
  const codes = value.split("").map((c) => int(c.charCodeAt(0)));

  return call(
    "new_string",
    call(
      "object",
      label("from", call("FixedArray", call("generics", "i32"), ...codes))
    )
  );
};

export const idIs = (id?: unknown, value: string) =>
  is(id, IdentifierAtom) && id.value === value;
