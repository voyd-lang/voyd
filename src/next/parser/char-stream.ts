import { SourceLocation } from "./ast/syntax.js";

export class CharStream {
  readonly filePath: string;
  readonly value: string;
  readonly length: number;
  readonly location = {
    index: 0,
    line: 1,
    column: 0,
  };

  constructor(contents: string, filePath: string) {
    this.value = contents;
    this.length = contents.length;
    this.filePath = filePath;
  }

  /** Current index the file is on */
  get position() {
    return this.location.index;
  }

  get line() {
    return this.location.line;
  }

  get column() {
    return this.location.column;
  }

  get hasCharacters() {
    return this.position < this.length;
  }

  get next() {
    return this.value[this.position];
  }

  currentSourceLocation() {
    const startLine = this.line;
    const startColumn = this.column;
    return new SourceLocation({
      startIndex: this.position,
      endIndex: this.position,
      startLine,
      endLine: startLine,
      startColumn,
      endColumn: startColumn,
      filePath: this.filePath,
    });
  }

  at(index: number): string | undefined {
    return this.value[this.position + index];
  }

  /** Returns the next character and advances the stream */
  consumeChar(): string {
    const char = this.next;
    if (char === undefined) {
      throw new Error("Out of characters");
    }

    this.location.index += 1;
    this.location.column += 1;
    if (char === "\n") {
      this.location.line += 1;
      this.location.column = 0;
    }

    return char;
  }
}
