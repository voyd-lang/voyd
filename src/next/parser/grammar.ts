import { Atom, IdentifierAtom } from "./ast/atom.js";
import { is, Syntax } from "./ast/syntax.js";

export const isIdentifier = (op?: unknown): op is IdentifierAtom =>
  is(op, IdentifierAtom);

export const identifierIs = (
  op: unknown | undefined,
  value: string
): op is Atom => isIdentifier(op) && op.value === value;

export const isOp = (op?: unknown): boolean => isInfixOp(op) || isPrefixOp(op);

/** Key is the operator, value is its precedence */
export type OpMap = Map<string, number>;

export const infixOps: OpMap = new Map([
  ["+", 1],
  ["-", 1],
  ["*", 2],
  ["/", 2],
  ["^", 3],
  ["%", 2],
  ["==", 2],
  ["!=", 2],
  ["<", 3],
  [">", 3],
  ["<=", 3],
  [">=", 3],
  [".", 6],
  ["|>", 4],
  ["<|", 4],
  ["|", 4],
  ["=", 0],
  ["+=", 4],
  ["-=", 4],
  ["*=", 4],
  ["/=", 4],
  ["->", 5],
  ["=>", 5],
  [":", 0],
  ["?:", 0],
  [":=", 0],
  ["::", 6],
  [";", 4],
  ["??", 3],
  ["?.", 6],
  ["and", 0],
  ["or", 0],
  ["xor", 0],
  ["as", 0],
  ["is", 0],
  ["is_subtype_of", 0],
  ["in", 0],
  ["has_trait", 0],
]);

const isUnquotedIdentifier = (op?: unknown): op is Atom =>
  isIdentifier(op) && !op.isQuoted;

export const isInfixOp = (op?: unknown): op is Atom =>
  isUnquotedIdentifier(op) && infixOps.has(op.value);

export const prefixOps: OpMap = new Map([
  ["#", 0],
  ["&", 7],
  ["!", 7],
  ["%", 7],
  ["$", 7],
  ["@", 7],
  ["$@", 7],
  ["...", 5],
]);

export const isPrefixOp = (op?: unknown): op is Atom =>
  isUnquotedIdentifier(op) && prefixOps.has(op.value);

export const greedyOps = new Set(["=>", "=", "<|", ";", "|"]);

export const isGreedyOp = (op?: unknown): op is Atom =>
  isUnquotedIdentifier(op) && greedyOps.has(op.value);

export const isContinuationOp = (op?: unknown) => isInfixOp(op);
