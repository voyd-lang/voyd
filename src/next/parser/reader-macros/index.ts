import { Token } from "../token.js";
import { arrayLiteralMacro } from "./array-literal.js";
import { booleanMacro } from "./boolean.js";
import { comment } from "./comment.js";
import { mapLiteralMacro } from "./map-literal.js";
import { floatMacro } from "./float.js";
import { intMacro } from "./int.js";
import { scientificENotationMacro } from "./scientific-e-notation.js";
import { stringMacro } from "./string.js";
import { objectLiteralMacro } from "./object-literal.js";
import { ReaderMacro } from "./types.js";
import { genericsMacro } from "./generics.js";
import { htmlMacro } from "./html/html.js";
import { parenReader } from "./paren.js";
import { whitespaceReader } from "./whitespace.js";
import { identifierReader } from "./identifier.js";
import { Expr } from "../ast/expr.js";

const MACROS = [
  parenReader,
  whitespaceReader,
  objectLiteralMacro,
  arrayLiteralMacro,
  mapLiteralMacro,
  intMacro,
  floatMacro,
  scientificENotationMacro,
  stringMacro,
  comment,
  booleanMacro,
  genericsMacro,
  htmlMacro,
  identifierReader,
];

export const getReaderMacroForToken = (
  token: Token,
  prev?: Expr,
  /** Next char */
  next?: string
): ReaderMacro["macro"] | undefined =>
  MACROS.find((m) => m.match(token, prev, next))?.macro;
