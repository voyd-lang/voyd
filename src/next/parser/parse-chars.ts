import { Expr, Form } from "./ast/index.js";
import { CharStream } from "./char-stream.js";
import { Lexer } from "./lexer.js";
import { getReaderMacroForToken } from "./reader-macros/index.js";
import { Token } from "./token.js";

export type ParseCharsOpts = {
  nested?: boolean;
  terminator?: string;
};

export const parseChars = (
  file: CharStream,
  opts: ParseCharsOpts = {}
): Form => {
  const lexer = new Lexer();
  const location = file.currentSourceLocation();
  const elements: Expr[] = [];

  while (file.hasCharacters) {
    const token = lexer.tokenize(file);

    if (token.value === opts.terminator) {
      break;
    }

    const result = processWithReaderMacro(token, file, elements.at(-1));
    if (result) elements.push(result);
  }

  location.setEndToStartOf(file.currentSourceLocation());
  const form = new Form({ location, elements });
  return opts.nested ? form : new Form(["ast", form]);
};

/** Returns true if token was matched with and processed by a macro  */
const processWithReaderMacro = (
  token: Token,
  file: CharStream,
  last?: Expr
) => {
  const readerMacro = getReaderMacroForToken(token, last, file.next);
  if (!readerMacro) return undefined;

  return readerMacro(file, {
    token,
    reader: (file, terminator) =>
      parseChars(file, {
        nested: true,
        terminator,
      }),
  });
};
