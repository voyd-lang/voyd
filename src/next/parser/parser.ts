import { Expr } from "./ast/expr.js";
import { CharStream } from "./char-stream.js";
import { read } from "./reader.js";
import { expandSyntaxMacros } from "./syntax-macros/index.js";

export const parse = (text: string, filePath?: string): Expr => {
  const chars = new CharStream(text, filePath ?? "raw");
  const rawAst = read(chars);
  return expandSyntaxMacros(rawAst);
};
