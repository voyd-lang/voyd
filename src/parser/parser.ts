import { List } from "../syntax-objects/index.js";
import { CharStream } from "./char-stream.js";
import { parseChars } from "./parse-chars.js";
import { ReaderMacro } from "./reader-macros/types.js";
import { expandSyntaxMacros } from "./syntax-macros/index.js";

export const parse = (text: string, filePath?: string): List => {
  const chars = new CharStream(text, filePath ?? "raw");
  const rawAst = parseChars(chars);
  return expandSyntaxMacros(rawAst);
};
