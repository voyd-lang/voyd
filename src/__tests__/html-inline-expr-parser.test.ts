import { describe, test, expect } from "vitest";
import { parse } from "../parser/parser.js";

describe("HTML inline expressions unwrap identifiers", () => {
  test("{item} is not wrapped as a list", () => {
    const text = `
use std::all
use std::vsx::create_element
use std::msg_pack::MsgPack

pub fn component()
  let a = ["Alex", "Abby"]
  <div>
    {a.map<MsgPack>((item) -> MsgPack => <p>hi {item}</p>)}
  </div>

pub fn main()
  msg_pack::encode(component())
`;

    const astJson = parse(text).toJSON();
    // Convert to plain JSON for easier traversal
    const plain = JSON.parse(JSON.stringify(astJson));

    // Walk the JSON to collect create_element calls
    const elements: unknown[] = [];
    const walk = (node: unknown) => {
      if (Array.isArray(node)) {
        if (node[0] === "create_element") elements.push(node);
        node.forEach(walk);
      } else if (node && typeof node === "object") {
        Object.values(node as Record<string, unknown>).forEach(walk);
      }
    };

    walk(plain);

    // Heuristic: inner <p> is the last create_element in this snippet
    const target = elements[elements.length - 1] as unknown[] | undefined;

    expect(target, "Found <p> element with item child").toBeTruthy();
    const children = (target![3] as unknown[]).slice(1);
    expect(children).toContain("item");
    expect(children).not.toContainEqual(["item"]);

    // Snapshot just the <p> element structure to lock in behavior
    expect(target).toMatchSnapshot();
  });
});
