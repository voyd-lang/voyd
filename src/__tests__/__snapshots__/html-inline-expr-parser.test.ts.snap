// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HTML inline expressions unwrap identifiers > {item} is not wrapped as a list 1`] = `
[
  "create_element",
  [
    "new_string",
    [
      "object",
      "ObjectLiteral-130",
      [
        [
          "from",
          [
            "FixedArray",
            112,
          ],
        ],
      ],
    ],
  ],
  [
    "array",
  ],
  [
    "array",
    [
      "new_string",
      [
        "object",
        "ObjectLiteral-144",
        [
          [
            "from",
            [
              "FixedArray",
              104,
              105,
              32,
            ],
          ],
        ],
      ],
    ],
    "item",
  ],
]
`;
